ibis/expr/rewrites.py

```diff
@@ -297,3 +297,80 @@ def simplify(node):
     node = node.replace(subsequent_projects | subsequent_filters)
     node = node.replace(complete_reprojection)
     return node
+
+
+@replace(p.Comparison)
+def rebase_predicates_on_table_underlying_jointable(_):
+    left, right = _.left, _.right
+    left_table = left.rel.parent
+    right_table = right.rel.parent
+    return _.copy(
+        left=left.copy(rel=left_table, name=left.name),
+        right=right.copy(rel=right_table, name=right.name),
+    )
+
+
+@replace(p.JoinChain)
+def rewrite_join_chain_for_semi_anti_join(_, y=None):
+    # TODO (mehmet): Added optional arg `y` to work around the
+    # error raised for `test_many_subqueries()`.
+    # Is there a cleaner way to address this?
+
+    links = _.find_topmost(ops.JoinLink)
+    semi_anti_links, new_links = [], []
+    for link in links:
+        if link.how in {"semi", "anti"}:
+            semi_anti_links.append(link)
+        else:
+            new_links.append(link)
+
+    # Nothing to do with the original join chain
+    if not semi_anti_links:
+        return _
+
+    if new_links:
+        node = _.copy(rest=tuple(new_links))
+    else:
+        node = _.first.parent
+
+    for link in semi_anti_links:
+        # Note: `filter` is necessary to avoid rebasing predicates
+        # in the `JoinChain`'s deeper down the tree. E.g., when
+        # a table is semi/anti-joined with an inner-join as
+        # table.join(inner_join, predicates=..., how="semi/anti")
+        link = link.replace(
```

Comment:
I assume this won't be required since the predicates are replaced below.

Reply:
It is required, otherwise we get the following error down the path:
```
    def test_semi_join_sql(semi_join_expr_and_expected_df, snapshot):
        expr, _ = semi_join_expr_and_expected_df
>       sql = ibis.to_sql(expr, dialect="flink")

ibis/backends/flink/tests/test_join.py:315:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ibis/expr/sql.py:385: in to_sql
    sg_expr = backend._to_sqlglot(expr.unbind(), **kwargs)
ibis/backends/flink/__init__.py:346: in _to_sqlglot
    return super()._to_sqlglot(expr, params=params)
ibis/backends/sql/__init__.py:160: in _to_sqlglot
    sql = self.compiler.translate(table_expr.op(), params=params)
ibis/backends/sql/compiler.py:525: in translate
    op, ctes = sqlize(
ibis/backends/sql/rewrites.py:284: in sqlize
    node = node.replace(reduce(operator.or_, rewrites))
ibis/common/graph.py:464: in replace
    results = self.map(replacer, filter=filter)
ibis/common/graph.py:265: in map
    results[node] = fn(node, results, **kwargs)
ibis/common/graph.py:188: in fn
    if (result := obj.match(recreated, ctx)) is NoMatch:
ibis/common/patterns.py:932: in match
    result = pattern.match(value, context)
ibis/common/patterns.py:932: in match
    result = pattern.match(value, context)
ibis/common/patterns.py:932: in match
    result = pattern.match(value, context)
ibis/common/patterns.py:932: in match
    result = pattern.match(value, context)
ibis/common/patterns.py:932: in match
    result = pattern.match(value, context)
ibis/common/patterns.py:391: in match
    return self.replacer.resolve(context)
ibis/common/deferred.py:326: in resolve
    return self.func(**context)
ibis/expr/rewrites.py:436: in rewrite_join_chain_for_semi_anti_join
    filtered = ops.Filter(parent=link.table.parent, predicates=predicates)
ibis/common/bases.py:72: in __call__
    return cls.__create__(*args, **kwargs)
ibis/common/grounds.py:120: in __create__
    return super().__create__(**kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ibis.expr.operations.relations.Filter object at 0x15e8f1e40>
parent = <ibis.expr.operations.temporal_windows.TumbleWindowingTVF object at 0x15e8f1c60>
predicates = (<ibis.expr.operations.logical.Equals object at 0x15eb18ac0>,)

    def __init__(self, parent, predicates):
        from ibis.expr.rewrites import ReductionLike

        for pred in predicates:
            if pred.find(ReductionLike, filter=Value):
                raise IntegrityError(
                    f"Cannot add {pred!r} to filter, it is a reduction which "
                    "must be converted to a scalar subquery first"
                )
            if pred.relations and parent not in pred.relations:
>               raise IntegrityError(
                    f"Cannot add {pred!r} to filter, they belong to another relation"
                )
E               ibis.common.exceptions.IntegrityError: Cannot add <ibis.expr.operations.logical.Equals object at 0x15eb18ac0> to filter, they belong to \
another relation
```